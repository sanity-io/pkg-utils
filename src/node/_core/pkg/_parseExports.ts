import {_isRecord} from '../_isRecord'
import {PkgExport} from '../config'
import {_PackageJSON} from './_types'
import {_validateExports} from './_validateExports'

/** @internal */
export function _parseExports(options: {pkg: _PackageJSON}): (PkgExport & {_path: string})[] {
  const {pkg} = options

  const rootExport: PkgExport & {_path: string} = {
    _exported: true,
    _path: '.',
    types: pkg.types,
    source: pkg.source || '',
    browser: pkg.browser && {
      source: pkg.source || '',
      require: pkg.main && pkg.browser[pkg.main],
      import: pkg.module && pkg.browser[pkg.module],
    },
    import: pkg.module,
    require: pkg.main,
    default: pkg.module || pkg.main || '',
  }

  const extraExports: (PkgExport & {_path: string})[] = []

  const errors: string[] = []

  if (pkg.exports) {
    if (!pkg.exports['./package.json']) {
      errors.push('package.json: `exports["./package.json"] must be declared.')
    }

    for (const [exportPath, exportEntry] of Object.entries(pkg.exports)) {
      if (exportPath.endsWith('.json')) {
        if (exportPath === './package.json') {
          if (exportEntry !== './package.json') {
            errors.push('package.json: `exports["./package.json"] must be "./package.json".')
          }
        }
      } else if (_isRecord(exportEntry)) {
        if (exportPath === '.') {
          if (
            exportEntry.require &&
            rootExport.require &&
            exportEntry.require !== rootExport.require
          ) {
            errors.push(
              'package.json: mismatch between "main" and "exports.require". These must be equal.'
            )
          }

          if (exportEntry.import && rootExport.import && exportEntry.import !== rootExport.import) {
            errors.push(
              'package.json: mismatch between "module" and "exports.import" These must be equal.'
            )
          }

          if (exportEntry.types && rootExport.types && exportEntry.types !== rootExport.types) {
            errors.push(
              'package.json: mismatch between "types" and "exports.types". These must be equal.'
            )
          }

          if (exportEntry.source && rootExport.source && exportEntry.source !== rootExport.source) {
            errors.push(
              'package.json: mismatch between "source" and "exports.source". These must be equal.'
            )
          }

          Object.assign(rootExport, exportEntry)
        } else {
          const extraExport: PkgExport & {_path: string} = {
            _exported: true,
            _path: exportPath,
            ...exportEntry,
          }

          extraExports.push(extraExport)
        }
      } else {
        errors.push('package.json: exports must be an object')
      }
    }
  }

  const _exports = [rootExport, ...extraExports]

  errors.push(..._validateExports(_exports, {pkg}))

  if (errors.length) {
    throw new Error('\n- ' + errors.join('\n- '))
  }

  return _exports
}
