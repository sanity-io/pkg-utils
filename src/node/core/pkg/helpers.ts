/** @internal */
export function assertFirst<T>(a: T, arr: T[]): boolean {
  const aIdx = arr.indexOf(a)

  // if not found, then we don't care
  if (aIdx === -1) {
    return true
  }

  return aIdx === 0
}

/** @internal */
export function assertLast<T>(a: T, arr: T[]): boolean {
  const aIdx = arr.indexOf(a)

  // if not found, then we don't care
  if (aIdx === -1) {
    return true
  }

  return aIdx === arr.length - 1
}

/** @internal */
export function assertOrder<T>(a: T, b: T, arr: T[]): boolean {
  const aIdx = arr.indexOf(a)
  const bIdx = arr.indexOf(b)

  // if either is not found, then we don't care
  if (aIdx === -1 || bIdx === -1) {
    return true
  }

  return aIdx < bIdx
}
