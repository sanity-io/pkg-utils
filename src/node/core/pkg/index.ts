export {loadPkg} from './loadPkg'
export {loadPkgWithReporting} from './loadPkgWithReporting'
export {parseExports} from './parseExports'
export {type PkgExtMap, pkgExtMap} from './pkgExt'
export type {PackageJSON} from './types'
