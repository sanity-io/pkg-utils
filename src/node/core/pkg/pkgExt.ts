import {cjsEnding, defaultEnding, legacyEnding, mjsEnding} from '../../tasks/dts/getTargetPaths'

/** @internal */
export interface PkgExtMap {
  commonjs: {commonjs: string; esm: string}
  module: {commonjs: string; esm: string}
  legacy: string
}

/** @internal */
export const pkgExtMap = {
  // pkg.type: "commonjs"
  commonjs: {
    commonjs: defaultEnding,
    esm: mjsEnding,
  },

  // pkg.type: "module"
  module: {
    commonjs: cjsEnding,
    esm: defaultEnding,
  },
  // package.config.legacyExports: true
  legacy: legacyEnding,
} satisfies PkgExtMap
