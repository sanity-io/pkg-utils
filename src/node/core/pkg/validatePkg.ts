import {z} from 'zod'

import type {PackageJSON} from './types'

const pkgSchema = z.object({
  type: z.optional(z.enum(['commonjs', 'module'])),
  name: z.string(),
  version: z.string(),
  license: z.string(),
  bin: z.optional(z.record(z.string())),
  dependencies: z.optional(z.record(z.string())),
  devDependencies: z.optional(z.record(z.string())),
  peerDependencies: z.optional(z.record(z.string())),
  source: z.optional(z.string()),
  main: z.optional(z.string()),
  browser: z.optional(z.record(z.string())),
  module: z.optional(z.string()),
  types: z.optional(z.string()),
  exports: z.optional(
    z.record(
      z.union([
        z.custom<`./${string}.json`>((val) => /^\.\/.*\.json$/.test(val as string)),
        z.custom<`./${string}.css`>((val) => /^\.\/.*\.css$/.test(val as string)),
        z.object({
          types: z.optional(z.string()),
          source: z.optional(z.string()),
          browser: z.optional(
            z.object({
              source: z.string(),
              import: z.optional(z.string()),
              require: z.optional(z.string()),
            }),
          ),
          node: z.optional(
            z.object({
              source: z.optional(z.string()),
              import: z.optional(z.string()),
              require: z.optional(z.string()),
            }),
          ),
          import: z.optional(z.string()),
          require: z.optional(z.string()),
          default: z.string(),
        }),
        z.object({
          types: z.optional(z.string()),
          svelte: z.string(),
          default: z.optional(z.string()),
        }),
      ]),
    ),
  ),
  browserslist: z.optional(z.union([z.string(), z.array(z.string())])),
  sideEffects: z.optional(z.union([z.boolean(), z.array(z.string())])),
  // @TODO type this properly
  typesVersions: z.optional(z.any()),
})

// Create a map over known keys to catch casing mistakes
const typoMap = new Map<string, string>()

for (const key of pkgSchema.keyof()._def.values) {
  typoMap.set(key.toUpperCase(), key)
}

export function validatePkg(input: unknown): PackageJSON {
  const pkg = pkgSchema.parse(input)

  const invalidKey = Object.keys(input as PackageJSON).find((key) => {
    const needle = key.toUpperCase()

    return typoMap.has(needle) ? typoMap.get(needle) !== key : false
  })

  if (invalidKey) {
    throw new TypeError(
      `
- package.json: "${invalidKey}" is not a valid key. Did you mean "${typoMap.get(invalidKey.toUpperCase())}"?`,
    )
  }

  return pkg
}
