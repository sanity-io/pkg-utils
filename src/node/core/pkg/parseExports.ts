import {existsSync} from 'node:fs'
import {resolve as resolvePath} from 'node:path'

import type {Logger} from '../../logger'
import type {InferredStrictOptions} from '../../strict'
import {defaultEnding, fileEnding, legacyEnding} from '../../tasks/dts/getTargetPaths'
import type {PkgExport} from '../config'
import {isRecord} from '../isRecord'
import {pkgExtMap} from './pkgExt'
import type {PackageJSON} from './types'
import {validateExports} from './validateExports'

/** @internal */
export function parseExports(options: {
  cwd: string
  pkg: PackageJSON
  strict: boolean
  strictOptions: InferredStrictOptions
  legacyExports: boolean
  logger: Logger
}): (PkgExport & {_path: string})[] {
  const {cwd, pkg, strict, strictOptions, legacyExports, logger} = options
  const type = pkg.type || 'commonjs'
  const errors: string[] = []

  const report = (kind: 'warn' | 'error', message: string) => {
    if (kind === 'warn') {
      logger.warn(message)
    } else {
      errors.push(message)
    }
  }

  if (!pkg.main && strict && strictOptions.alwaysPackageJsonMain !== 'off') {
    report(strictOptions.alwaysPackageJsonMain, 'package.json: `main` must be declared')
  }

  if (!Array.isArray(pkg.files) && strict && strictOptions.alwaysPackageJsonFiles !== 'off') {
    report(
      strictOptions.alwaysPackageJsonFiles,
      'package.json: `files` should be used over `.npmignore`',
    )
  }

  if (pkg.source) {
    if (
      strict &&
      pkg.exports?.['.'] &&
      typeof pkg.exports['.'] === 'object' &&
      'source' in pkg.exports['.'] &&
      pkg.exports['.'].source === pkg.source
    ) {
      errors.push(
        'package.json: the "source" property can be removed, as it is equal to exports["."].source.',
      )
    } else if (!pkg.exports && pkg.main) {
      const extMap = pkgExtMap[type]
      const importExport = pkg.main.replace(fileEnding, extMap.esm)
      const requireExport = pkg.main.replace(fileEnding, extMap.commonjs)
      const defaultExport = pkg.main.replace(fileEnding, defaultEnding)

      const maybeBrowserCondition = []

      if (pkg.browser) {
        const browserConditions = []

        if (pkg.module && pkg.browser?.[pkg.module]) {
          browserConditions.push(
            `      "import": ${JSON.stringify(pkg.browser[pkg.module].replace(fileEnding, extMap.esm))}`,
          )
        } else if (pkg.browser?.[pkg.main]) {
          browserConditions.push(
            `      "import": ${JSON.stringify(pkg.browser?.[pkg.main].replace(fileEnding, extMap.esm))}`,
          )
        } else if (legacyExports) {
          const browserImport = pkg.main.replace(fileEnding, `.browser${extMap.esm}`)

          browserConditions.push(`      "import": ${JSON.stringify(browserImport)}`)
        }

        if (pkg.browser?.[pkg.main]) {
          browserConditions.push(
            `      "require": ${JSON.stringify(pkg.browser[pkg.main].replace(fileEnding, extMap.commonjs))}`,
          )
        } else if (legacyExports) {
          const browserRequire = pkg.main.replace(fileEnding, `.browser${extMap.commonjs}`)

          browserConditions.push(`      "require": ${JSON.stringify(browserRequire)}`)
        }

        if (browserConditions.length) {
          maybeBrowserCondition.push(
            `    "browser": {`,
            `      "source": ${JSON.stringify(pkg.browser?.[pkg.source] || pkg.source)},`,
            ...browserConditions,
            `    }`,
          )
        }
      }

      errors.push(
        ...([
          'package.json: `exports` are missing, it should be:',
          `"exports": {`,
          `  ".": {`,
          `    "source": ${JSON.stringify(pkg.source)},`,
          // If browser conditions are detected then add them to the suggestion
          ...(maybeBrowserCondition.length > 0 ? maybeBrowserCondition : []),
          // If legacy exports are enabled we suggest the full list of exports, if not we can use the terse version
          (legacyExports || type === 'commonjs') &&
            `    "import": ${JSON.stringify(importExport)},`,
          (legacyExports || type === 'module') &&
            `    "require": ${JSON.stringify(requireExport)},`,
          `    "default": ${JSON.stringify(defaultExport)}`,
          `  },`,
          `  "./package.json": "./package.json"`,
          `}`,
        ].filter(Boolean) as string[]),
      )
    }
  }

  if (errors.length) {
    throw new Error('\n- ' + errors.join('\n- '))
  }

  if (!pkg.exports) {
    throw new Error(
      '\n- ' +
        [
          'package.json: `exports` are missing, please set a minimal configuration, for example:',
          `"exports": {`,
          `  ".": {`,
          `    "source": "./src/index.js",`,
          `    "default": "./dist/index.js"`,
          `  },`,
          `  "./package.json": "./package.json"`,
          `}`,
        ].join('\n- '),
    )
  }

  const _exports: (PkgExport & {_path: string})[] = []

  // @TODO validate typesVersions when legacyExports is true

  if (strict && strictOptions.noPackageJsonTypings !== 'off' && 'typings' in pkg) {
    report(strictOptions.noPackageJsonTypings, 'package.json: `typings` should be `types`')
  }

  if (
    strict &&
    strictOptions.alwaysPackageJsonTypes !== 'off' &&
    !pkg.types &&
    typeof pkg.exports?.['.'] === 'object' &&
    'source' in pkg.exports['.'] &&
    pkg.exports['.'].source?.endsWith('.ts')
  ) {
    report(
      strictOptions.alwaysPackageJsonTypes,
      'package.json: `types` must be declared for the npm listing to show as a TypeScript module.',
    )
  }

  if (strict && !pkg.exports['./package.json']) {
    errors.push('package.json: `exports["./package.json"] must be declared.')
  }

  for (const [exportPath, exportEntry] of Object.entries(pkg.exports)) {
    if (
      exportPath.endsWith('.json') ||
      (typeof exportEntry === 'string' && exportEntry.endsWith('.json'))
    ) {
      if (exportPath === './package.json') {
        if (exportEntry !== './package.json') {
          errors.push('package.json: `exports["./package.json"]` must be "./package.json".')
        }
      }
    } else if (exportPath.endsWith('.css')) {
      if (typeof exportEntry === 'string' && !existsSync(resolvePath(cwd, exportEntry))) {
        errors.push(
          `package.json: \`exports[${JSON.stringify(exportPath)}]\`: file does not exist.`,
        )
      } else if (typeof exportEntry !== 'string') {
        errors.push(
          `package.json: \`exports[${JSON.stringify(exportPath)}]\`: export conditions not supported for CSS files.`,
        )
      }
    } else if (isRecord(exportEntry) && 'svelte' in exportEntry) {
      // @TODO should we report a warning or a debug message here about a detected svelte export that is ignored?
    } else if (isPkgExport(exportEntry)) {
      const exp = {
        _exported: true,
        _path: exportPath,
        ...exportEntry,
      } satisfies PkgExport & {_path: string}

      // Infer the `default` condition based on the `type` and other conditions
      if (!exp.default) {
        const fallback = type === 'module' ? exp.import : exp.require

        if (fallback) {
          exp.default = fallback
        }

        if (legacyExports) {
          if (fallback) {
            errors.push(
              `package.json - \`exports["${exp._path}"].default\` should be set to "${fallback}" when "legacyExports" is true`,
            )
          } else {
            errors.push(
              `package.json - \`exports["${exp._path}"].default\` should be specified when "legacyExports" is true`,
            )
          }
        }
      }

      // Infer the `require` condition based on the `type` and other conditions
      if (!exp.require && type === 'commonjs' && exp.default) {
        exp.require = exp.default
      }

      // Infer the `import` condition based on the `type` and other conditions
      if (!exp.import && type === 'module' && exp.default) {
        exp.import = exp.default
      }

      if (exportPath === '.') {
        if (exportEntry.require && pkg.main && exportEntry.require !== pkg.main) {
          errors.push(
            'package.json: mismatch between "main" and "exports.require". These must be equal.',
          )
        }

        if (legacyExports) {
          const indexLegacyExport = (exportEntry.import || exportEntry.require || '').replace(
            /(\.esm)?\.[mc]?js$/,
            legacyEnding,
          )

          if (indexLegacyExport !== pkg.module) {
            errors.push(
              `package.json: "module" should be "${indexLegacyExport}" when "legacyExports" is true`,
            )
          }
        } else {
          if (exportEntry.import && pkg.module && exportEntry.import !== pkg.module) {
            errors.push(
              'package.json: mismatch between "module" and "exports.import" These must be equal.',
            )
          }
        }
      }

      _exports.push(exp)
    } else if (!isRecord(exportEntry)) {
      errors.push('package.json: exports must be an object')
    }
  }

  errors.push(...validateExports(_exports, {pkg}))

  if (errors.length) {
    throw new Error('\n- ' + errors.join('\n- '))
  }

  return _exports
}

function isPkgExport(value: unknown): value is PkgExport {
  return isRecord(value) && 'source' in value && typeof value['source'] === 'string'
}
