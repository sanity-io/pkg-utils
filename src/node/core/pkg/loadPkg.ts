import fs from 'node:fs/promises'

import pkgUp from 'pkg-up'

import type {PackageJSON} from './types'
import {validatePkg} from './validatePkg'

/** @internal */
export async function loadPkg(options: {cwd: string}): Promise<PackageJSON> {
  const {cwd} = options

  const pkgPath = await pkgUp({cwd})

  if (!pkgPath) throw new Error('no package.json found')

  const buf = await fs.readFile(pkgPath)

  const raw = JSON.parse(buf.toString())

  validatePkg(raw)

  return raw
}
