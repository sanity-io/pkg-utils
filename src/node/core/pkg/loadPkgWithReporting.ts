import chalk from 'chalk'
import {ZodError} from 'zod'

import type {Logger} from '../../logger'
import {assertLast, assertOrder} from './helpers'
import {loadPkg} from './loadPkg'
import type {PackageJSON} from './types'

/** @internal */
export async function loadPkgWithReporting(options: {
  cwd: string
  logger: Logger
  strict: boolean
  legacyExports: boolean
}): Promise<PackageJSON> {
  const {cwd, logger, strict, legacyExports} = options

  try {
    const pkg = await loadPkg({cwd})
    let shouldError = false

    if (strict) {
      if (legacyExports && pkg.type === 'commonjs') {
        shouldError = true
        logger.error(
          `the \`type\` field in \`./package.json\` shouldn't be "commonjs" when \`legacyExports\` is set to true)`,
        )
      }

      if (!legacyExports && !pkg.type) {
        shouldError = true
        logger.error(
          `the \`type\` field in \`./package.json\` must be either "module" or "commonjs")`,
        )
      }
    }

    // validate exports
    if (pkg.exports) {
      const _exports = Object.entries(pkg.exports)

      for (const [expPath, exp] of _exports) {
        if (typeof exp === 'string' || 'svelte' in exp) {
          continue
        }

        const keys = Object.keys(exp)

        if (exp.types) {
          shouldError = true
          logger.error(
            `exports["${expPath}"]: the \`types\` condition shouldn't be used as dts files are generated in such a way that both CJS and ESM is supported`,
          )
        }

        if (exp.module) {
          shouldError = true
          logger.error(
            `exports["${expPath}"]: the \`module\` condition shouldn't be used as it's not well supported in all bundlers.`,
          )
        }

        if (exp.node) {
          if (exp.import && exp.node.import && !assertOrder('node', 'import', keys)) {
            shouldError = true
            logger.error(
              `exports["${expPath}"]: the \`node\` property should come before the \`import\` property`,
            )
          }

          if (exp.node.module) {
            shouldError = true
            logger.error(
              `exports["${expPath}"]: the \`node.module\` condition shouldn't be used as it's not well supported in all bundlers. A better strategy is to refactor the codebase to no longer be vulnerable to the "dual package hazard"`,
            )
          }

          if (
            !exp.node.source &&
            exp.node.import &&
            (exp.node.require || exp.require) &&
            (exp.node.import.endsWith('.cjs.js') || exp.node.import.endsWith('.cjs.mjs'))
          ) {
            shouldError = true
            logger.error(
              `exports["${expPath}"]: the \`node.import\` re-export pattern shouldn't be used as it's not well supported in all bundlers. A better strategy is to refactor the codebase to no longer be vulnerable to the "dual package hazard"`,
            )
          }

          if (exp.require && exp.node.require && exp.require === exp.node.require) {
            shouldError = true
            logger.error(
              `exports["${expPath}"]: the \`node.require\` property isn't necessary as it's identical to \`require\``,
            )
          } else if (exp.require && exp.node.require && !assertOrder('node', 'require', keys)) {
            shouldError = true
            logger.error(
              `exports["${expPath}"]: the \`node\` property should come before the \`require\` property`,
            )
          }
        } else {
          if (!assertOrder('import', 'require', keys)) {
            logger.warn(
              `exports["${expPath}"]: the \`import\` property should come before the \`require\` property`,
            )
          }
        }

        if (!assertLast('default', keys)) {
          shouldError = true
          logger.error(
            `exports["${expPath}"]: the \`default\` property should be the last property`,
          )
        }
      }
    }

    if (shouldError) {
      process.exit(1)
    }

    return pkg
  } catch (err) {
    if (err instanceof ZodError) {
      for (const issue of err.issues) {
        if (issue.code === 'invalid_type') {
          logger.error(
            [
              `\`${formatPath(issue.path)}\` `,
              `in \`./package.json\` must be of type ${chalk.magenta(issue.expected)} `,
              `(received ${chalk.magenta(issue.received)})`,
            ].join(''),
          )
          continue
        }

        logger.error(issue)
      }
    } else {
      logger.error(err)
    }

    process.exit(1)
  }
}

function formatPath(segments: Array<string | number>) {
  return segments
    .map((s, idx) => {
      if (idx === 0) return s

      if (typeof s === 'number') {
        return `[${s}]`
      }

      if (s.startsWith('.')) {
        return `["${s}"]`
      }

      return `.${s}`
    })
    .join('')
}
