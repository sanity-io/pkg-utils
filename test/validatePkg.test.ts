import {expect, test} from 'vitest'

import {validatePkg} from '../src/node/core/pkg/validatePkg'

const template = {
  type: 'module',
  name: 'dummy-module',
  version: '0.0.0',
  license: 'MIT',
  exports: {
    './package.json': './package.json',
  },
}

test.each([
  {actual: 'Exports' as const, expected: 'exports', value: {'./package.json': './package.json'}},
  {actual: 'Type', expected: 'type', value: 'module'},
  {actual: 'Dependencies', expected: 'dependencies', value: {}},
  {actual: 'devdependencies', expected: 'devDependencies', value: {}},
  {actual: 'peerdependencies', expected: 'peerDependencies', value: {}},
  {actual: 'Source', value: './src/index.ts'},
  {actual: 'Main', value: './dist/index.js'},
  {actual: 'Browser', value: {}},
  {actual: 'Module', value: './dist/index.js'},
  {actual: 'Types', value: './dist/index.js'},
  {
    actual: 'typesversions',
    expected: 'typesVersions',
    value: {
      '*': {
        extra: ['./dist/extra.d.ts'],
      },
    },
  },
])('$actual is not a valid key, did you mean $expected?', ({actual, expected, value}) => {
  const pkg = {
    ...template,
    [actual]: value,
  }

  expect(() => validatePkg(pkg)).toThrowError(/is not a valid key/)
  expect(() => validatePkg(pkg)).toThrowErrorMatchingSnapshot()
  expect(() => validatePkg({...template, [expected as string]: value})).not.toThrow()
})

test('conditional imports must be prefixed with #', () => {
  expect(() =>
    validatePkg({
      ...template,
      imports: {
        env: {
          browser: './src/index.browser.ts',
          default: './src/index.ts',
        },
      },
    }),
  ).toThrowErrorMatchingSnapshot()
  expect(
    validatePkg({
      ...template,
      imports: {
        '#env': {
          browser: './src/index.browser.ts',
          default: './src/index.ts',
        },
      },
    }),
  ).toMatchSnapshot()
})
