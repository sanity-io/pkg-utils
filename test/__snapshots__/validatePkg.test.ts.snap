// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`'Browser' is not a valid key, did you mean undefined? 1`] = `
[TypeError: 
- package.json: "Browser" is not a valid key. Did you mean "browser"?]
`;

exports[`'Dependencies' is not a valid key, did you mean 'dependencies'? 1`] = `
[TypeError: 
- package.json: "Dependencies" is not a valid key. Did you mean "dependencies"?]
`;

exports[`'Exports' is not a valid key, did you mean 'exports'? 1`] = `
[TypeError: 
- package.json: "Exports" is not a valid key. Did you mean "exports"?]
`;

exports[`'Main' is not a valid key, did you mean undefined? 1`] = `
[TypeError: 
- package.json: "Main" is not a valid key. Did you mean "main"?]
`;

exports[`'Module' is not a valid key, did you mean undefined? 1`] = `
[TypeError: 
- package.json: "Module" is not a valid key. Did you mean "module"?]
`;

exports[`'Source' is not a valid key, did you mean undefined? 1`] = `
[TypeError: 
- package.json: "Source" is not a valid key. Did you mean "source"?]
`;

exports[`'Type' is not a valid key, did you mean 'type'? 1`] = `
[TypeError: 
- package.json: "Type" is not a valid key. Did you mean "type"?]
`;

exports[`'Types' is not a valid key, did you mean undefined? 1`] = `
[TypeError: 
- package.json: "Types" is not a valid key. Did you mean "types"?]
`;

exports[`'devdependencies' is not a valid key, did you mean 'devDependencies'? 1`] = `
[TypeError: 
- package.json: "devdependencies" is not a valid key. Did you mean "devDependencies"?]
`;

exports[`'peerdependencies' is not a valid key, did you mean 'peerDependencies'? 1`] = `
[TypeError: 
- package.json: "peerdependencies" is not a valid key. Did you mean "peerDependencies"?]
`;

exports[`'typesversions' is not a valid key, did you mean 'typesVersions'? 1`] = `
[TypeError: 
- package.json: "typesversions" is not a valid key. Did you mean "typesVersions"?]
`;

exports[`conditional imports must be prefixed with # 1`] = `
[ZodError: [
  {
    "code": "custom",
    "fatal": true,
    "path": [
      "imports",
      "env"
    ],
    "message": "Invalid input"
  }
]]
`;

exports[`conditional imports must be prefixed with # 2`] = `
{
  "exports": {
    "./package.json": "./package.json",
  },
  "imports": {
    "#env": {
      "browser": "./src/index.browser.ts",
      "default": "./src/index.ts",
    },
  },
  "license": "MIT",
  "name": "dummy-module",
  "type": "module",
  "version": "0.0.0",
}
`;
